### PR TITLE
WIP - Alternate config formats (Do Not Merge) See: #691

### PR DESCRIPTION
**Pull request by jgandt**
This is just a Q&D proof of concept to show the parsing mechanism.

I have not written tests. I have not squashed commits. Lots of other stuff was left out so that we can initiate a conversation about the mechanism and architecture.

### CODE REVIEW

File: lib/tmuxinator/project.rb [hunk 1]
```diff
@@ -41,33 +41,17 @@ class Project
...
-        args = options[:args] || []
-        @settings = parse_settings(args)
-        @args = args
-
-        content = Erubis::Eruby.new(raw_content).result(binding)
-        YAML.safe_load(content, [], [], true)
+        config_file_type = options[:type] || 'default'
+        # now must pass in type
+        # I temporarily hard coded it for getting 'procfile' type up and running
+        yml = Parser.new(path, options, 'procfile').build_yaml
```
Comment (jgandt) [inline lib/tmuxinator/project.rb]: Obviously the 'procfile' type is temporarily hard coded. I didn't want to dive into cli changes until we get the parsing mechanism close to what we want.

**Comment by jgandt** [comment]
Also this is not 100% working for the Procfile, so I still have some work to do there. I'm not entirely sure what I did wrong, but the yaml I created in `Parsers::Procfile` doesn't quite do the trick.

**Comment by ethagnawl** [comment]
Hey, @jgandt. 

This is a great start and I'm definitely interested in seeing this fleshed out.

A few quick thoughts:
- As this stands, it looks this approach will coerce any/all config files _into_ YAML. Right? This makes sense as a starting point and/or bridge to the final implementation, but my initial thought was that we'd have a config file class/hash/similar which provides the uniform interface to the config data. Granted, that would probably complicate the implementation of new config file types, but would prevent tmuxinator from arbitrarily leaning on YAML. I could be swayed, though, so I'd definitely be interested in your thoughts. (One complication with my approach is with how ERB interpolation would be handled -- I'll need to think some more about how the parsers would handle that and maybe look at how projects like Sprockets handle that (e.g. some-library.js.erb, which first interpolates any Ruby code contained in the JS file).)
- Following up on that, I think YAML should lose its special status and be treated as Just Another Parser Class (:stuck_out_tongue_winking_eye:). So, `Default` would become `YAML`.

Any thoughts on the above, @tmuxinator/tmuxinator?

**Comment by jgandt** [comment]
So I'm parsing it into "yaml." It's really just a ruby representation of the current config file format. That is, YAML can be directly translated into ruby objects. Therefore the current config file format already provides the uniform interface to the config data. So really I'm just converting the Procfile into that pre-defined interface. I am happy to wrap a class or code around it so we can provide some better documentation for that format. That being said, I believe you already have some documentation around valid configuration files in the form of examples. However if we still want to codify it in the code, I'm happy to so.

It just so happens that the Procfile format is parseable by YAML parsers. The format is:
```
process_name_1: process command 1
process_name_2: process command 2
...
```
So I'm really just using a quirk of the format to conveniently parse the file. Then I stuff it into the "yaml" variable. I'll rename some variables and methods and we can see if that makes things clearer.

**Comment by jgandt** [comment]
> IMO, this would be great. It'd be nice to have a defined API and would make it simple for others to introduce new parsers.

You got it! I'll try to put it together tonight.

**Comment by jgandt** [comment]
Ok I think I'm going to do some refactoring in the project. I'd like to reclaim the `Config` module to handle the actual configuration object that will come out of parsed files.

Right now, the `Config` module is really more of a config file finder. I will try to slowly refactor the code to reclaim the module if that is ok.

If you would prefer to leave it as is, I can find another name for the config object. `ProjectConfig` or something would work. It might even be more clear on the intent.

I took a small first shot at some refactoring that I think is a good one even if we don't rename the module. see: #700.

**Comment by adamstrickland** [comment]
> It'd be nice to have a defined API and would make it simple for others to introduce new parsers. 

I think that this is a possible long-term goal.  However, maybe before we start down that road, we wait to see what kind of demand we're looking at.  I have created https://github.com/tmuxinator/tmuxinator/issues/703 to solicit community feedback on a more general solution.

I think I would prefer incorporating *just* the Procfile parsing with minimal refactoring as the right step at this time.

**Pull request closed without merging**